### PR TITLE
Run linter on the interop branch

### PR DESCRIPTION
**Pull request by jklaise**
This should fix flake8 and mypy issues. Also a few bugs with docs fixed. Merge when checks pass.

### CODE REVIEW

File: alibi/explainers/cf_example_wb.py [hunk 1]
```diff
@@ -1,10 +1,12 @@
+# flake8 noqa
 import tensorflow as tf
```
Comment (alexcoca) [inline alibi/explainers/cf_example_wb.py:2]: This file and its `*bb.py` counterpart were used by myself to do some sanity checks when making big changes. Do we keep them in this branch until we finish the development or shall we remove them?

File: alibi/explainers/experimental/counterfactuals.py [hunk 2]
```diff
@@ -110,32 +109,32 @@
...
-    `lam_exploration_steps` consecutive iteration. The computational time is reduced as a result.
-    - `instance_proba_delta`: the model prediction on the counterfactual for the class with the highest probability for \
-    the original instance should decrease by at least this value in `lam_exploration_steps` to warrant continuation of \
-    optimisation for a given :math:`\lambda` in the initial search loop
-    - `lam_cf_threshold`, `lam_multiplier`, `lam_divider`: see `bisect_lam` docstring in `Watcher counterfactual`
-    - `common_ration`, `decay_steps`: role in determining the original optimisation schedule for :math:`\lambda` as  \
-    described above
-    - 'max_lam_steps': maximum number of times to adjust the regularization constant before terminating the search \
-    (number of outer loops).
+ which depends on the hyperparameter :math:`\lambda`. The explainer will first try to determine a suitable range for
```
Comment (alexcoca) [inline alibi/explainers/experimental/counterfactuals.py:112]: Have we checked this renders fine?

File: alibi/utils/frameworks.py [hunk 3]
```diff
@@ -69,9 +70,9 @@ def infer_device(predictor, predictor_type: str, framework: str) -> Union[None,
     """
 
     if framework == 'tensorflow':
-        return  # type: ignore
+        return None
     if predictor_type == 'blackbox':
-        return
+        return None
```
Comment (alexcoca) [inline alibi/utils/frameworks.py:75]: I disagree with this. `return` means `return None` in `python`, this is a language convention. If the typing library doesn't know about this, then this is a problem with it so I say we should stick with the `# type:  ignore` directive.

File: alibi/explainers/experimental/counterfactuals.py [hunk 4]
```diff
@@ -1062,8 +1061,8 @@ def __init__(self,
             'method_opts': method_opts,
             'feature_range': feature_range,
             'framework': framework,
-        }
-        self._explainer = self._explainer_type(*explainer_args, **explainer_kwargs, **kwargs)
+        }  # type: Dict
+        self._explainer = self._explainer_type(*explainer_args, **explainer_kwargs, **kwargs)  # type: ignore
```
Comment (alexcoca) [inline alibi/explainers/experimental/counterfactuals.py:1065]: Why do we ingnore the type here? Isn't this a `dict`?

File: alibi/utils/frameworks.py [hunk 5]
```diff
@@ -36,12 +37,12 @@ def tensorflow_installed() -> bool:
     if not has_tensorflow:
         raise ImportError(template.format(pkg=tf_required))
     if int(tf_version[0]) > 1:
```
Comment (alexcoca) [inline alibi/utils/frameworks.py:39]: We need to introduce here the change we discussed privately, otherwise the docs do not build.

**Review by alexcoca — CHANGES_REQUESTED** [review]
Hi Janis,

Thanks for cleaning this up. We need to add appropriate casting in `alibi.utils.frameworks.py` as we discussed, otherwise the docs won't build. Other than that just one or two comments about typing.

File: alibi/explainers/cf_example_wb.py (hunk 1 above)
Comment (jklaise) [reply alibi/explainers/cf_example_wb.py:2]: Can keep them as long as we remember to delete before merge to master.

File: alibi/explainers/experimental/counterfactuals.py (hunk 2 above)
Comment (jklaise) [reply alibi/explainers/experimental/counterfactuals.py:112]: I checked it.

File: alibi/utils/frameworks.py (hunk 3 above)
Comment (jklaise) [reply alibi/utils/frameworks.py:75]: Ok fair enough, I just thought this is cleaner/shorter than doing the extra type declaration.

File: alibi/explainers/experimental/counterfactuals.py (hunk 4 above)
Comment (jklaise) [reply alibi/explainers/experimental/counterfactuals.py:1065]: There was a reason for this...

Comment (jklaise) [reply alibi/explainers/experimental/counterfactuals.py:1065]: With no type info I get the following error: `alibi/explainers/experimental/counterfactuals.py:1065: error: Function gets multiple values for keyword argument "predictor"`

Comment (jklaise) [reply alibi/explainers/experimental/counterfactuals.py:1065]: It's because `predictor` is somehow present in both `explainer_args` and `kwargs`.

Comment (jklaise) [reply alibi/explainers/experimental/counterfactuals.py:1065]: The correct type (correct me if I'm wrong) would be `#type: _WachterCounterfactual`, but the problem persists, so I just originally ignored the type.

Comment (jklaise) [reply alibi/explainers/experimental/counterfactuals.py:1065]: This is a known bug so I would stick with `# type: ignore`: https://github.com/python/mypy/issues/6799

Comment (alexcoca) [reply alibi/explainers/experimental/counterfactuals.py:1065]: Agreed.

File: alibi/utils/frameworks.py (hunk 3 above)
Comment (alexcoca) [reply alibi/utils/frameworks.py]: I still insist :)

Comment (alexcoca) [reply alibi/utils/frameworks.py]: Actually, I take back my original assertion. Having researched a bit, we can see that your modification makes sense in this case, according to [this](https://stackoverflow.com/questions/15300550/return-return-none-and-no-return-at-all) source.

**Review by alexcoca — APPROVED** [review]
All nice and tiy, I'll merge this.

**Pull request merged**